gnuradio3_{9,8}: fix build

chuangzhu opened the pull request:
## Description of changes

gnuradio3_9: add the unreleased upstream commit that fix build with GCC 13

gnuradio3_8: use GCC 12 with libstdc++ from GCC 13

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

doronbehar commented:
Hello and thanks for observing the issue and fixing it! Changes look pretty good to me, though I noticed there's a `pkgs/by-name` CI error I don't fully understand, and I would like the commit messages to be:

```
gnuradio3_9Minimal: fix compilation with gcc 13
gnuradio3_8Minimal: fix build by using gcc 12
```

chuangzhu committed:
gnuradio3_9Minimal: fix compilation with gcc 13

chuangzhu committed:
gnuradio3_8Minimal: fix build on Linux by using gcc 12 libraries

Not using simply stdenv = gcc12Stdenv due to errors like this:

/nix/store/...-binutils-2.40/bin/ld: /nix/store/...-boost-1.81.0/lib/libboost_program_options.so.1.81.0: undefined reference to `std::ios_base_library_init()@GLIBCXX_3.4.32'
collect2: error: ld returned 1 exit status
make[2]: *** [gr-fec/lib/reed-solomon/CMakeFiles/gr_fec_rstest.dir/build.make:131: gr-fec/lib/reed-solomon/gr_fec_rstest] Error 1
make[1]: *** [CMakeFiles/Makefile2:2165: gr-fec/lib/reed-solomon/CMakeFiles/gr_fec_rstest.dir/all] Error 2

chuangzhu commented:
We have never been using `pkgs/by-name` for gnuradio, so it should be a bug of the workflow. Even if we are going to migrate to `pkgs/by-name`, it should be in another PR. I guess we should just ignore the check.

doronbehar commented:
Cc @infinisil for this `pkgs/by-name` error.

@chuangzhu Any idea how come the darwin builds fail now in CI, and they didn't fail in https://github.com/NixOS/nixpkgs/pull/284482 ?

On a different topic, I'm investigating locally if upstream issue https://github.com/gnuradio/gnuradio/issues/3801 is indeed fixed.

doronbehar commented on pkgs/top-level/all-packages.nix:
```diff
@@ -31666,12 +31666,13 @@ with pkgs;
       };
     };
   };
-  gnuradio3_8 = disable-warnings-if-gcc13 (callPackage ../applications/radio/gnuradio/wrapper.nix {
```
In my local testings I noticed that if `disable-warnings-if-gcc13` would have been applied to the unwrapped derivation, it would have worked. @chuangzhu do you have strong preference to using gcc12 vs disabling gcc13 warnings?

chuangzhu replied:
Interesting, it failed in my local test https://gist.github.com/chuangzhu/658ab70024fea247959c5f00aa914fee

doronbehar replied:
> Interesting, it failed in my local test [gist.github.com/chuangzhu/658ab70024fea247959c5f00aa914fee](https://gist.github.com/chuangzhu/658ab70024fea247959c5f00aa914fee)

I was mistaken, sorry about that (confused GR3.9 and GR3.8)...

Even if I'll revert this to your version of the PR, I wonder how can we fix the Darwin build errors.. According to [this successful hydra build](
https://nydra.nixos.org/build/247654168), and it's nixpkgs revision, the Darwin build didn't fail there thanks to `disable-warnings-if-gcc13` (I also verified the default `gcc --version` is 13.2.0 there. I wonder, why did you use `stdenvAdapters.useLibsFrom`, and not simply `gcc12Stdenv`?

chuangzhu replied:
I guess Darwin uses LLVM so it's not affected by the GCC update. For Linux however, there seems to be some ABI incompatibility between boost and GCC 12's libstdc++

doronbehar commented:
> On a different topic, I'm investigating locally if upstream issue [gnuradio/gnuradio#3801](https://github.com/gnuradio/gnuradio/issues/3801) is indeed fixed.

Not fixed...

doronbehar commented:
I'm pushing my version of this PR where the `disable-warnings-if-gcc13` is applied to the unwrapped derivation - to test if it fixes this new darwin failure.

infinisil commented:
> We have never been using `pkgs/by-name` for gnuradio, so it should be a bug of the workflow. Even if we were going to migrate to `pkgs/by-name`, it should be in another PR. I guess we should just ignore the check.

Indeed, sorry for that! Will fix this week

doronbehar commented:
I pushed an attempt to fix both Darwin and Linux builds.

doronbehar commented:
I will squash all the `gnuradio3_8Minimal:` commits when testing will be done.

infinisil commented:
Oh actually this `pkgs/by-name` check failure should be fixed now that I merged https://github.com/NixOS/nixpkgs/pull/284765. So the next time you push to this PR, `pkgs/by-name` CI should turn green

doronbehar commented:
I'm pasting here the errors ofborg got, for reference:

```
/nix/store/...-binutils-2.40/bin/ld: /nix/store/...-boost-1.81.0/lib/libboost_program_options.so.1.81.0: undefined reference to `std::ios_base_library_init()@GLIBCXX_3.4.32'
collect2: error: ld returned 1 exit status
make[2]: *** [gr-fec/lib/reed-solomon/CMakeFiles/gr_fec_rstest.dir/build.make:131: gr-fec/lib/reed-solomon/gr_fec_rstest] Error 1
make[1]: *** [CMakeFiles/Makefile2:2165: gr-fec/lib/reed-solomon/CMakeFiles/gr_fec_rstest.dir/all] Error 2
```

I also put these in the commit message for later reference.

doronbehar commented:
Pushed another attempt.

doronbehar commented:
OK CI is green. Apparently the `disable-warnings-if-gcc13` is redundant if gcc12 is used - I checked for a difference with `nix derivation show` before and after. I removed it, and I'll merge after evaluation is finished.
